### ISSUE TITLE
Blocking await triggers on original message

### ISSUE THREAD

**Issue by HorizonShadow** [meew0/discordrb#513]
Using `await!` on a user will sometimes trigger on the command message.

You can see it in this example. `!!afk` registers the await, and after it triggers, it prints `Await received, id: #{message}`

You can see that sometimes it works correct, and sometimes it will trigger on the `!!afk` message itself.
```
[8:02 PM] Cereal: !!afk
[8:02 PM] Cereal: e
[8:02 PM] BOTyuela-test-bot: Await recieved, id: e
[8:02 PM] Cereal: !!afk
[8:02 PM] BOTyuela-test-bot: Await recieved, id: !!afk
```

Code for this command:

```ruby
begin
  p "Command run, id: #{e.message}"
  response = e.user.await!
  p "Await recieved, id: #{response.message}"
rescue StandardError => e
  e
end
```

**Comment by z64** [comment]
Please show the event handler you are using (i.e. `bot.message`) and your `Bot.new` line

**Comment by HorizonShadow** [comment]
```ruby
BOT = Discordrb::Commands::CommandBot.new({
  token: CONFIG['discord'],
  prefix: '!!',
  log_level: :debug
})
```

The event handler is a *little* quirky, sorry:

```ruby
Commands.constants.map do |c|
  command = Commands.const_get(c)
  command.is_a?(Class) ? command : nil
end.compact.each do |command|
  BOT.command(command.name, command.attributes, &command.command)
end
```

and then the command is

```ruby
module Commands
  class AfkCommand
    class << self
      def name
        :afk
      end

      def attributes
        {
          min_args: 0,
          usage: 'afk [message?]',
          description: 'Set yourself as afk'
        }
      end

      def command
        lambda do |e, *message|
          begin
            p "Command run, id: #{e.message}"
            response = e.message.await!
            p "Await recieved, id: #{response.message}"
          rescue StandardError => e
            e
          end
        end
      end
    end
  end
end
```

**Comment by z64** [comment]
I can't reproduce this:

```
[9:56 PM] z64: !!test-user
[9:56 PM] BOTdahdev: What is your name?
[9:56 PM] z64: foo
[9:56 PM] BOTdahdev: Hello, foo!
[9:57 PM] z64: (shouldn't respond to this message)

[9:57 PM] z64: !!test-message
[9:57 PM] BOTdahdev: What is your name?
[9:57 PM] z64: bar
[9:57 PM] BOTdahdev: Hello, bar!
[9:57 PM] z64: (shouldn't respond to this message)

[9:58 PM] z64: !!test-command
[9:58 PM] BOTdahdev: What is your name?
[9:58 PM] z64: baz
[9:58 PM] BOTdahdev: Hello, baz!
[9:58 PM] z64: (shouldn't respond to this message)
```
```rb
require "discordrb"

bot = Discordrb::Commands::CommandBot.new token: ENV['TOKEN'], prefix: "!!"

bot.message(content: "!!test-user", in: "#debug") do |event|
  event.respond "What is your name?"
  response = event.user.await!
  event.respond "Hello, #{response.message}!"
end

bot.message(content: "!!test-message", in: "#debug") do |event|
  event.respond "What is your name?"
  response = event.message.await!
  event.respond "Hello, #{response.message}!"
end

bot.command(:"test-command") do |event|
  event.respond "What is your name?"
  response = event.message.await!
  event.respond "Hello, #{response.message}!"
end

bot.run
```

`await!` simply wraps basic event handlers. `event.user.await!` is the same exact thing as:
```rb
bot.message(content: "!!test-user", in; "debug") do |event|
  new_handler = bot.message(from: event.user.id) do |inner_event|
    inner_event.respond "boop!"
  end

  event.respond "Created handler: #{new_handler}, will be removed in 3 seconds"
  sleep 3
  bot.remove_handler(new_handler)
  event.respond "Stopped!"
end
```

Except, its wrapped in a Thread, Mutex and CV that deletes the temporary event handler after receiving an event.

References:

- [Message#await!](https://github.com/meew0/discordrb/blob/master/lib/discordrb/data.rb#L2528-L2532)
- [User#await!](https://github.com/meew0/discordrb/blob/master/lib/discordrb/data.rb#L241-L246)
- [Bot#add_await!](https://github.com/meew0/discordrb/blob/master/lib/discordrb/bot.rb#L563-L598)

(note how the deprecated `await` registers events in the same exact way)

I'll need more information to debug this further. Other users have been using `await!` as well without reporting anything like this, so I'm really not sure.

**Comment by HorizonShadow** [comment]
Can you try running an empty message handler, followed by the command containing the await!?

For example:

```ruby
BOT.message do |event|

end

BOT.command(:whatever) do |event|
  await!
  p event.message.content
end
```

I was commented out bits of the bot until I couldn't reproduce the problem anymore and it appears to be related to a catch all handler I have running. If I comment it out, I stop running into the problem.

**Comment by z64** [comment]
Ah okay, I understand why this happens now. I'll do my best to explain.

## 1. How event handlers are stored internally

Internally we store `@event_handlers`, which is `Hash<EventClass, Array<EventHandler>>`

So doing:

```rb
bot.message { }
```

Creates in `@event_handlers`:

```rb
{Discordrb::Events::MessageEvent=>
  [#<Discordrb::Events::MessageEventHandler:0x000055c67dc2f198
    @attributes={},
    @block=#<Proc:0x000055c67dc2f760@issue-513.rb:6>>]}
```

## 2. How events are dispatched

There are two systems at play here. *Command dispatch* and what I'll call *regular event dispatch*.

See [source](https://github.com/meew0/discordrb/blob/master/lib/discordrb/bot.rb#L994-L1023), as I explain this:

When we receive Discord's `MESSAGE_CREATE`, we first do some checks that it isn't ourself (the bot) or a user we are ignoring.

Next, we call the `create_message` internal event handler which *is a method that does nothing* by default.

This `create_message` method is overwritten by `CommandBot` to process and raise `CommandEvents`. See [source](https://github.com/meew0/discordrb/blob/master/lib/discordrb/commands/command_bot.rb#L362-L393)

The important thing to understand here is that **commands will always be processed before event handlers**.

So *after* this, we go into dispatching across our *regular* event handlers.

## 3. Raising events

See [source](https://github.com/meew0/discordrb/blob/master/lib/discordrb/bot.rb#L1268-L1277)

```rb
    def raise_event(event)
      debug("Raised a #{event.class}")
      handle_awaits(event)

      @event_handlers ||= {}
      handlers = @event_handlers[event.class]
      (handlers || []).each do |handler|
        call_event(handler, event) if handler.matches?(event)
      end
    end
```

As I explained before, all that `add_await!` does is register a `MessageEventHandler` *just* like `bot.message` does (so you can ignore `handle_awaits`). So what this does is iterate over the event handlers and `call_event` if the event `matches?` the event's attributes (who the message is from, etc, any other filters)

To summarize the "race condition" that is occurring:

1. We get `MESSAGE_CREATE` from Discord.

2. `CommandBot` gets the message first, and will spawn a new thread to handle the message in `execute_chain`.

3. The bots internals continue on (with the command running in the background) to `raise_event` to call any event handlers that have been registered.

4. If you have no event handlers, `@event_handlers` is `nil` and no work is done very quickly. (`[].each` is performed)

5. If you have event handlers, and it matches, it will be called in its own thread per-handler. Creating a thread takes time. So what is happening, in the space of time it took to create a new thread and process your "empty" event handler *first*, your command thread has been running in the background and `add_await!` has created an event handler that is added to `@event_handlers` *as it is being iterated over*, causing the same message that ran your command to also trigger the await's event handler that got added in "late".

I hope this makes sense. You're right it's an issue, and I have some ideas on how it could be fixed.

**Comment by z64** [comment]
Oh, and if you want a "silly" workaround :smile: :heart: 

```rb
bot.command(:whatever) do |event|
  Discordrb::LOGGER.info "Setting up await for: #{event.user.distinct}"
  loop do
    response = event.user.await!
    # Only break if we got a different message, if not, try again:
    break if response.message.id != event.message.id
  end
  "command done"
end
```

**Comment by HorizonShadow** [comment]
Yup that makes sense.

Thanks so much for the detailed explanation, too!

**Comment by z64** [comment]
No problem, thanks for testing.

**Comment by z64** [comment]
Hi @HorizonShadow 

Please see #514 . I think this will fix this issue with minimal performance impact. If you have time, please feel free to use my branch to test it or monkey patch the internal method with the same diff.

I have tested it briefly on a small bot like the ones we have discussed here and it seems to work fine, though this was a difficult bug to reproduce in the first place :smile:

**Comment by z64** [comment]
I've merged this now in c8109e5. If anyone still experiences issues or has problems because of that patch, feel free to comment here or open a new issue :+1:
